File: app/src/main/java/br/com/example/hamburgos/model/Snack.java
package br.com.example.hamburgos.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class Snack implements Serializable {

    private int id;
    private String image;
    private String name;

    private List<Ingredient> ingredientList;
    private List<Ingredient> extras;

    public int getId() {
        return id;
    }

    public String getImage() {
        return image;
    }

    public String getName() {
        return name;
    }

    public float getPrice(){
        float result = 0;
        if ( ingredientList != null ) {
            for (Ingredient ingredient : ingredientList) {
                result += ingredient.getPrice();
            }
        }
        if ( extras != null) {
            for (Ingredient ingredient : extras) {
                result += ingredient.getPrice();
            }
        }


        if ( isLight() ){
            result = result * (float)0.90;
        }
        return result;
    }

    private boolean isLight(){
        boolean result = false;
        boolean lettuceFlag = checkIfIngredientInList("Alface");
        if ( lettuceFlag ){
            result = !checkIfIngredientInList("Bacon");
        }
        return result;
    }

    private boolean checkIfIngredientInList(String ingredientName){
        boolean result = false;
        if ( ingredientName!= null ) {
            if (ingredientList != null) {
                for (Ingredient ingredient : ingredientList) {
                    if (ingredientName.toLowerCase().equals(ingredient.getName().toLowerCase())) {
                        result = true;
                        break;
                    }
                }
            }
            if (!result && extras != null) {
                for (Ingredient ingredient : extras) {
                    if (ingredientName.toLowerCase().equals(ingredient.getName().toLowerCase())) {
                        result = true;
                        break;
                    }
                }
            }
        }
        return result;
    }

    public List<Ingredient> getIngredientList() {
        return ingredientList;
    }

    public String getIngredientListString(){
        String result = "";
        if ( ingredientList != null ) {
            for (Ingredient ingredient : ingredientList) {
                result += ingredient.getName() + ", ";
            }
        }
        if ( extras != null) {
            for (Ingredient ingredient : extras) {
                result += ingredient.getName() + ", ";
            }
        }
        if ( result.length() > 2) {
            result = result.substring(0, result.length() - 2);
        }
        return result;
    }

    public void setIngredientList(List<Ingredient> ingredientList) {
        this.ingredientList = ingredientList;
    }

    public List<Ingredient> getExtras() {
        return extras;
    }

    public void setExtras(List<Ingredient> extras) {
        this.extras = extras;
    }

    public void addExtras(Ingredient ingredient){
        if (extras == null ){
            extras = new ArrayList<>();
        }
        extras.add(ingredient);
    }

    public void removeExtras(Ingredient ingredient) {
        if ( extras != null ) {
            extras.remove(ingredient);
        }
    }
}

File: app/src/main/java/br/com/example/hamburgos/model/Snack.kt
package br.com.example.hamburgos.model

import br.com.example.hamburgos.util.Constants
import java.io.Serializable
import java.util.*

class Snack : Serializable {

    val id: Int = 0
    val image: String? = null
    val name: String = ""

    var ingredientList: List<Ingredient> = listOf()
    var extras: MutableList<Ingredient> = mutableListOf()

    val price: Float
        get() {
            var result = 0f
            for (ingredient in ingredientList) {
                result += ingredient.price
            }

            for (ingredient in extras) {
                result += ingredient.price
            }


            val meat = howManyIngredientOnList(Constants.HAMBURGER)
            if (meat > 2 && meat % 3 == 0) {
                result -= (meat * 3).toFloat()
                result += (meat / 3 * 2 * 3).toFloat()
            }

            val cheese = howManyIngredientOnList(Constants.CHEESE)
            if (cheese > 2 && cheese % 3 == 0) {
                result -= (cheese * 1.5).toFloat()
                result += (cheese / 3 * 2 * 1.5).toFloat()
            }

            if (isLight) {
                result *= 0.90.toFloat()
            }
            return result
        }

    private val isLight: Boolean
        get() {
            var result = false
            val lettuceFlag = checkIfIngredientInList(Constants.LETTUCE)
            if (lettuceFlag) {
                result = !checkIfIngredientInList("Bacon")
            }
            return result
        }

    val ingredientListString: String
        get() {
            var result = ""

            for (ingredient in ingredientList) {
                result += ingredient.name
                result = "$result, "
            }

            for (ingredient in extras) {
                result += ingredient.name
                result = "$result, "
            }
            if (result.length > 2) {
                result = result.substring(0, result.length - 2)
            }
            return result
        }

    val jsonExtras: List<Int>
        get() {
            return extras.map { id }
        }

    private fun checkIfIngredientInList(ingredientName: String): Boolean {
        for (ingredient in ingredientList) {
            if (ingredientName.toLowerCase() == ingredient.name.toLowerCase()) {
                return true
            }

        }
        for (ingredient in extras) {
            if (ingredientName.toLowerCase() == ingredient.name.toLowerCase()) {
                return true
            }
        }
        return false
    }

    private fun howManyIngredientOnList(ingredientName: String): Int {
        var result = 0
        for (ingredient in ingredientList) {
            if (ingredientName.toLowerCase() == ingredient.name.toLowerCase()) {
                ++result
            }
        }
        for (ingredient in extras) {
            if (ingredientName.toLowerCase() == ingredient.name.toLowerCase()) {
                ++result
            }
        }
        return result
    }

    fun addExtras(ingredient: Ingredient) {
        extras.add(ingredient)
    }

    fun removeExtras(ingredient: Ingredient) {
        extras.remove(ingredient)
    }
}
